# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: security/sandbox/chromium/sandbox/win/src/sid_unittest.cc
# Commit: 370c40e49230
# Full Hash: 370c40e4923073a87845b57c1f4851bacd9663da
# Author: Bob Owen <bobowencode@gmail.com>
# Date: 2019-06-12 21:42:35
# Regressor Bug: 1552160
# File Overlap Count: 3
# Description:
#   Bug 1552160 Part 1: Roll-up of chromium sandbox update and mozilla patches to get a running browser. r=jld,aklotz,tjr,bobowen
#   
#   This updates security/sandbox/chromium/ files to chromium commit 84108231f6e6e0772fb9a4643679ce76aa771e67.
#   
#   Existing and new patches applied from security/sandbox/chromium-shim/patches/with_update/ to give a compiling and mostly working browser.
# ==============================================================================

diff -r bd5ecd72f76c -r 370c40e49230 security/sandbox/chromium/sandbox/win/src/sid_unittest.cc
--- a/security/sandbox/chromium/sandbox/win/src/sid_unittest.cc	Wed Jun 12 21:55:59 2019 +1200
+++ b/security/sandbox/chromium/sandbox/win/src/sid_unittest.cc	Wed Jun 12 11:10:48 2019 +0100
@@ -4,68 +4,179 @@
 
 // This file contains unit tests for the sid class.
 
-#define _ATL_NO_EXCEPTIONS
-#include <atlbase.h>
-#include <atlsecurity.h>
+#include "sandbox/win/src/sid.h"
 
-#include "sandbox/win/src/sid.h"
+#include <sddl.h>
+
+#include "base/win/atl.h"
+#include "base/win/windows_version.h"
 #include "testing/gtest/include/gtest/gtest.h"
 
 namespace sandbox {
 
-// Calls ::EqualSid. This function exists only to simplify the calls to
-// ::EqualSid by removing the need to cast the input params.
-BOOL EqualSid(const SID *sid1, const SID *sid2) {
-  return ::EqualSid(const_cast<SID*>(sid1), const_cast<SID*>(sid2));
+namespace {
+
+bool EqualSid(const Sid& sid, const ATL::CSid& compare_sid) {
+  if (!sid.IsValid())
+    return false;
+  return !!::EqualSid(sid.GetPSID(), const_cast<SID*>(compare_sid.GetPSID()));
 }
 
-// Tests the creation if a Sid
+bool EqualSid(const Sid& sid, const wchar_t* sddl_sid) {
+  PSID compare_sid;
+  if (!sid.IsValid())
+    return false;
+  if (!::ConvertStringSidToSid(sddl_sid, &compare_sid))
+    return false;
+  bool equal = !!::EqualSid(sid.GetPSID(), compare_sid);
+  ::LocalFree(compare_sid);
+  return equal;
+}
+
+struct KnownCapabilityTestEntry {
+  WellKnownCapabilities capability;
+  const wchar_t* sddl_sid;
+};
+
+struct NamedCapabilityTestEntry {
+  const wchar_t* capability_name;
+  const wchar_t* sddl_sid;
+};
+
+}  // namespace
+
+// Tests the creation of a Sid.
 TEST(SidTest, Constructors) {
   ATL::CSid sid_world = ATL::Sids::World();
-  SID *sid_world_pointer = const_cast<SID*>(sid_world.GetPSID());
+  PSID sid_world_pointer = const_cast<SID*>(sid_world.GetPSID());
 
-  // Check the SID* constructor
+  // Check the SID* constructor.
   Sid sid_sid_star(sid_world_pointer);
-  ASSERT_TRUE(EqualSid(sid_world_pointer, sid_sid_star.GetPSID()));
+  ASSERT_TRUE(EqualSid(sid_sid_star, sid_world));
+
+  // Check the copy constructor.
+  Sid sid_copy(sid_sid_star);
+  ASSERT_TRUE(EqualSid(sid_copy, sid_world));
 
-  // Check the copy constructor
-  Sid sid_copy(sid_sid_star);
-  ASSERT_TRUE(EqualSid(sid_world_pointer, sid_copy.GetPSID()));
+  Sid sid_sddl = Sid::FromSddlString(L"S-1-1-0");
+  ASSERT_TRUE(sid_sddl.IsValid());
+  ASSERT_TRUE(EqualSid(sid_sddl, sid_world));
+
+  Sid sid_sddl_invalid = Sid::FromSddlString(L"X-1-1-0");
+  ASSERT_FALSE(sid_sddl_invalid.IsValid());
+
+  Sid sid_sddl_empty = Sid::FromSddlString(L"");
+  ASSERT_FALSE(sid_sddl_empty.IsValid());
 
   // Note that the WELL_KNOWN_SID_TYPE constructor is tested in the GetPSID
-  // test.
+  // test. AppContainer related constructors are tested in AppContainer.
 }
 
 // Tests the method GetPSID
 TEST(SidTest, GetPSID) {
   // Check for non-null result;
-  ASSERT_NE(static_cast<SID*>(NULL), Sid(::WinLocalSid).GetPSID());
-  ASSERT_NE(static_cast<SID*>(NULL), Sid(::WinCreatorOwnerSid).GetPSID());
-  ASSERT_NE(static_cast<SID*>(NULL), Sid(::WinBatchSid).GetPSID());
+  ASSERT_NE(nullptr, Sid(::WinLocalSid).GetPSID());
+  ASSERT_NE(nullptr, Sid(::WinCreatorOwnerSid).GetPSID());
+  ASSERT_NE(nullptr, Sid(::WinBatchSid).GetPSID());
+
+  ASSERT_TRUE(EqualSid(Sid(::WinNullSid), ATL::Sids::Null()));
+
+  ASSERT_TRUE(EqualSid(Sid(::WinWorldSid), ATL::Sids::World()));
+
+  ASSERT_TRUE(EqualSid(Sid(::WinDialupSid), ATL::Sids::Dialup()));
+
+  ASSERT_TRUE(EqualSid(Sid(::WinNetworkSid), ATL::Sids::Network()));
+
+  ASSERT_TRUE(
+      EqualSid(Sid(::WinBuiltinAdministratorsSid), ATL::Sids::Admins()));
+
+  ASSERT_TRUE(EqualSid(Sid(::WinBuiltinUsersSid), ATL::Sids::Users()));
+
+  ASSERT_TRUE(EqualSid(Sid(::WinBuiltinGuestsSid), ATL::Sids::Guests()));
+
+  ASSERT_TRUE(EqualSid(Sid(::WinProxySid), ATL::Sids::Proxy()));
+}
 
-  ASSERT_TRUE(EqualSid(Sid(::WinNullSid).GetPSID(),
-                       ATL::Sids::Null().GetPSID()));
+TEST(SidTest, KnownCapability) {
+  if (base::win::GetVersion() < base::win::VERSION_WIN8)
+    return;
+
+  Sid sid_invalid_well_known =
+      Sid::FromKnownCapability(kMaxWellKnownCapability);
+  EXPECT_FALSE(sid_invalid_well_known.IsValid());
 
-  ASSERT_TRUE(EqualSid(Sid(::WinWorldSid).GetPSID(),
-                       ATL::Sids::World().GetPSID()));
-
-  ASSERT_TRUE(EqualSid(Sid(::WinDialupSid).GetPSID(),
-                       ATL::Sids::Dialup().GetPSID()));
+  const KnownCapabilityTestEntry capabilities[] = {
+      {kInternetClient, L"S-1-15-3-1"},
+      {kInternetClientServer, L"S-1-15-3-2"},
+      {kPrivateNetworkClientServer, L"S-1-15-3-3"},
+      {kPicturesLibrary, L"S-1-15-3-4"},
+      {kVideosLibrary, L"S-1-15-3-5"},
+      {kMusicLibrary, L"S-1-15-3-6"},
+      {kDocumentsLibrary, L"S-1-15-3-7"},
+      {kEnterpriseAuthentication, L"S-1-15-3-8"},
+      {kSharedUserCertificates, L"S-1-15-3-9"},
+      {kRemovableStorage, L"S-1-15-3-10"},
+      {kAppointments, L"S-1-15-3-11"},
+      {kContacts, L"S-1-15-3-12"},
+  };
 
-  ASSERT_TRUE(EqualSid(Sid(::WinNetworkSid).GetPSID(),
-                       ATL::Sids::Network().GetPSID()));
+  for (auto capability : capabilities) {
+    EXPECT_TRUE(EqualSid(Sid::FromKnownCapability(capability.capability),
+                         capability.sddl_sid))
+        << "Known Capability: " << capability.sddl_sid;
+  }
+}
+
+TEST(SidTest, NamedCapability) {
+  if (base::win::GetVersion() < base::win::VERSION_WIN10)
+    return;
+
+  Sid sid_nullptr = Sid::FromNamedCapability(nullptr);
+  EXPECT_FALSE(sid_nullptr.IsValid());
 
-  ASSERT_TRUE(EqualSid(Sid(::WinBuiltinAdministratorsSid).GetPSID(),
-                       ATL::Sids::Admins().GetPSID()));
+  Sid sid_empty = Sid::FromNamedCapability(L"");
+  EXPECT_FALSE(sid_empty.IsValid());
+
+  const NamedCapabilityTestEntry capabilities[] = {
+      {L"internetClient", L"S-1-15-3-1"},
+      {L"internetClientServer", L"S-1-15-3-2"},
+      {L"registryRead",
+       L"S-1-15-3-1024-1065365936-1281604716-3511738428-"
+       "1654721687-432734479-3232135806-4053264122-3456934681"},
+      {L"lpacCryptoServices",
+       L"S-1-15-3-1024-3203351429-2120443784-2872670797-"
+       "1918958302-2829055647-4275794519-765664414-2751773334"},
+      {L"enterpriseAuthentication", L"S-1-15-3-8"},
+      {L"privateNetworkClientServer", L"S-1-15-3-3"}};
 
-  ASSERT_TRUE(EqualSid(Sid(::WinBuiltinUsersSid).GetPSID(),
-                       ATL::Sids::Users().GetPSID()));
+  for (auto capability : capabilities) {
+    EXPECT_TRUE(EqualSid(Sid::FromNamedCapability(capability.capability_name),
+                         capability.sddl_sid))
+        << "Named Capability: " << capability.sddl_sid;
+  }
+}
 
-  ASSERT_TRUE(EqualSid(Sid(::WinBuiltinGuestsSid).GetPSID(),
-                       ATL::Sids::Guests().GetPSID()));
+TEST(SidTest, Sddl) {
+  Sid sid_sddl = Sid::FromSddlString(L"S-1-1-0");
+  ASSERT_TRUE(sid_sddl.IsValid());
+  base::string16 sddl_str;
+  ASSERT_TRUE(sid_sddl.ToSddlString(&sddl_str));
+  ASSERT_EQ(L"S-1-1-0", sddl_str);
+}
 
-  ASSERT_TRUE(EqualSid(Sid(::WinProxySid).GetPSID(),
-                       ATL::Sids::Proxy().GetPSID()));
+TEST(SidTest, SubAuthorities) {
+  DWORD world_subauthorities[] = {0};
+  SID_IDENTIFIER_AUTHORITY world_authority = {SECURITY_WORLD_SID_AUTHORITY};
+  Sid sid_world =
+      Sid::FromSubAuthorities(&world_authority, 1, world_subauthorities);
+  ASSERT_TRUE(EqualSid(sid_world, ATL::Sids::World()));
+  ASSERT_TRUE(Sid::FromSubAuthorities(&world_authority, 0, nullptr).IsValid());
+
+  DWORD admin_subauthorities[] = {32, 544};
+  SID_IDENTIFIER_AUTHORITY nt_authority = {SECURITY_NT_AUTHORITY};
+  Sid sid_admin =
+      Sid::FromSubAuthorities(&nt_authority, 2, admin_subauthorities);
+  ASSERT_TRUE(EqualSid(sid_admin, ATL::Sids::Admins()));
 }
 
 }  // namespace sandbox